`r/azurerm_managed_disk`: Add support for `hyper_v_generation` property

myc2h6o opened the pull request:
Fixes #9137

- The property doesn't support update in Azure go sdk thus setting `ForceNew` to true.
- Test result:
![image](https://user-images.githubusercontent.com/10579712/144357894-285afe54-0db1-4019-b01c-7d68c4863e1d.png)

katbyte commented on internal/services/compute/managed_disk_resource.go:
```diff
@@ -187,6 +187,17 @@ func resourceManagedDisk() *pluginsdk.Resource {
 				ValidateFunc: validation.IntBetween(2, 10),
 			},
 
+			"hyper_v_generation": {
```
lets leave version on it and maybe add a note to that other file to rename it to be more clear - thanks!

myc2h6o replied:
sure, will use hyper_v_generation_version instead and rename the property in other files

tombuildsstuff replied:
generation and version mean essentially the same thing in this context - but this is known as the Generation of HyperV, rather than Version - so we should use `generation` and not `version` here

katbyte commented:
@myc2h6o - any update on renaming the property?

myc2h6o commented:
@katbyte i'll work on the renaming today, was adding other properties earlier

myc2h6o commented:
@katbyte I updated the property name to `hyper_v_geneartion_version`, and left a note in other three occurrences of `hyper_v_geneartion` saying rename in v3.0. Besides this, I fixed an issue in the code as well to not include it in the request if the property is empty, please take another look.

As for renaming the property, will it be taken care of together with other breaking changes in the 3.0 release?

tombuildsstuff commented on internal/services/compute/managed_disk_resource.go:
```diff
@@ -215,6 +215,17 @@ func resourceManagedDisk() *pluginsdk.Resource {
 				ForceNew: true,
 			},
 
+			"hyper_v_generation_version": {
```
'generation' implies this is a version - so the `version` suffix here is unnecessary:

```suggestion
			"hyper_v_generation": {
```

myc2h6o replied:
_version removed

tombuildsstuff commented on internal/services/compute/managed_disk_resource.go:
```diff
@@ -215,6 +215,17 @@ func resourceManagedDisk() *pluginsdk.Resource {
+			"hyper_v_generation_version": {
+				Type:     pluginsdk.TypeString,
+				Optional: true,
+				Computed: true,
```
whilst this comes from the image, since this is the disk resource we should make users either specify this, or use ignore_changes on it:

```suggestion
```

myc2h6o replied:
Computed removed

tombuildsstuff commented on internal/services/compute/managed_disk_resource.go:
```diff
@@ -812,6 +827,7 @@ func resourceManagedDiskRead(d *pluginsdk.ResourceData, meta interface{}) error
 		d.Set("os_type", props.OsType)
 		d.Set("tier", props.Tier)
 		d.Set("max_shares", props.MaxShares)
+		d.Set("hyper_v_generation_version", props.HyperVGeneration)
```
is there a default value for this for older disk images? iirc there wasn't - so this'll be null & should be defaulted in the read function is the user doesn't specify a value here

myc2h6o replied:
The default value when not specified is an empty string (not nil) as the type of props.HyperVGeneration is not a pointer.

tombuildsstuff commented on internal/services/compute/image_resource.go:
```diff
@@ -54,6 +54,7 @@ func resourceImage() *pluginsdk.Resource {
 				ForceNew: true,
 			},
 
+			// TODO: rename to hyper_v_generation_version in v3.0
```
(as below) we can remove this, the name is fine:

```suggestion
```

myc2h6o replied:
Removed

tombuildsstuff commented on internal/services/compute/shared_image_data_source.go:
```diff
@@ -49,6 +49,7 @@ func dataSourceSharedImage() *pluginsdk.Resource {
 				Computed: true,
 			},
 
+			// TODO: rename to hyper_v_generation_version in v3.0
```
(as above) we can remove this, the name is fine:

```suggestion
```

tombuildsstuff commented on internal/services/compute/shared_image_resource.go:
```diff
@@ -67,6 +67,7 @@ func resourceSharedImage() *pluginsdk.Resource {
 				}, false),
 			},
 
+			// TODO: rename to hyper_v_generation_version in v3.0
```
(as above) we can remove this, the name is fine:

```suggestion
```

tombuildsstuff commented on website/docs/r/managed_disk.html.markdown:
```diff
@@ -111,6 +111,8 @@ The following arguments are supported:
 
 * `encryption_settings` - (Optional) A `encryption_settings` block as defined below.
 
+* `hyper_v_generation_version` - (Optional) The generation of HyperV of the Disk. Applicable to OS disks only. Possible values are `V1` and `V2`. Changing this forces a new resource to be created.
```
```suggestion
* `hyper_v_generation` - (Optional) The HyperV Generation of the Disk, which can only be set for OS Disks. Possible values are `V1` and `V2`. Changing this forces a new resource to be created.
```

myc2h6o replied:
Doc updated

myc2h6o replied:
I added a similar clarification like in `os_type`. This property is similar to `os_type`, which can be specified by user when an image reference is not provided, but will be assigned a value by server from the image if it is specified.
I'm thinking if we need to add more comments to this property as well as `os_type` explaining when to use `ignore_changes` for these two properties. What do you think @tombuildsstuff

tombuildsstuff reviewed (changes requested):
hey @myc2h6o 

Thanks for this PR.

I've taken a look through and left some comments inline, but if we can remove the `_version` suffix here (since it's superfluous) then we should otherwise be good to go 👍 

Thanks!

katbyte reviewed (approved):
LGTM 💠
